src/som/interpreter/nodes/literals/GenericArrayLiteralNode.java

```diff
@@ -0,0 +1,186 @@
+    super(source);
+    assert source != null;
+    this.expressions = expressions.toArray(new ExpressionNode[0]);
+  }
+
+  private boolean allComponentsAreBoolean(final Object[] arr) {
+    for (Object a : arr) {
+      if (!(a instanceof Boolean)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean allComponentsAreLong(final Object[] arr) {
+    for (Object a : arr) {
+      if (!(a instanceof Long)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean allComponentsAreDouble(final Object[] arr) {
+    for (Object a : arr) {
+      if (!(a instanceof Double)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean allComponentsAreString(final Object[] arr) {
+    for (Object a : arr) {
+      if (!(a instanceof String)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @ExplodeLoop
+  protected SMutableArray evaluatedExpressions(final VirtualFrame frame) {
+    Object[] arr = new Object[expressions.length];
+    for (int i = 0; i < expressions.length; i++) {
+      arr[i] = expressions[i].executeGeneric(frame);
+    }
+
+    GenericArrayLiteralNode replacedNode = this;
+
+    if (allComponentsAreBoolean(arr)) {
+      replacedNode = replace(new BooleanArrayLiteralNode(Arrays.asList(expressions), sourceSection));
+    } else if (allComponentsAreLong(arr)) {
+      replacedNode = replace(new LongArrayLiteralNode(Arrays.asList(expressions), sourceSection));
+    } else if (allComponentsAreDouble(arr)) {
+      replacedNode = replace(new DoubleArrayLiteralNode(Arrays.asList(expressions), sourceSection));
+    } else if (allComponentsAreString(arr)) {
+      replacedNode = replace(new StringArrayLiteralNode(Arrays.asList(expressions), sourceSection));
+    }
+
+    if (replacedNode != this) {
+      return replacedNode.evaluatedExpressions(frame);
+    } else {
+      return new SMutableArray(arr, Classes.arrayClass);
+    }
+
+  }
+
+  @Override
+  public SArray executeSArray(final VirtualFrame frame) throws UnexpectedResultException {
+    return evaluatedExpressions(frame);
+  }
+
+  @Override
+  public Object executeGeneric(final VirtualFrame frame) {
+    return evaluatedExpressions(frame);
+  }
+
+  public final class BooleanArrayLiteralNode extends GenericArrayLiteralNode {
+
+    public BooleanArrayLiteralNode(final List<ExpressionNode> expressions, final SourceSection source) {
+      super(expressions, source);
+    }
+
+    @Override
+    @ExplodeLoop
+    protected SMutableArray evaluatedExpressions(final VirtualFrame frame) {
+      boolean[] arr = new boolean[expressions.length];
+      for (int i = 0; i < expressions.length; i++) {
+        try {
+          arr[i] = expressions[i].executeBoolean(frame);
+        } catch (UnexpectedResultException e) {
+          throw new RuntimeException("Expression " + i + " of this BooleanArrayLiteralNode did not evaluate to a double");
```

Comment:
Ehm, ok?
That's not a semantically correct way to handle it :)
You can cheat, as long as you don't get caught :)